compiler/src/dotty/tools/dotc/core/Types.scala

```diff
@@ -3012,15 +3012,19 @@ object Types {
 
     override def superType(implicit ctx: Context): Type = {
       if (ctx.period != validSuper) {
+        var canCache = true
         cachedSuper = tycon match {
```

Comment:
Modifying `cachedSuper` when we cannot cache the value seems dangerous. If `validSuper` refers to a non-empty period, then a subsequent call to `superType` might be done with `ctx.period` equal to `validSuper` and would return the new, incorrect value in `cachedSuper`.